img2dataset/downloader.py

```diff
@@ -34,6 +36,69 @@ def is_disallowed(headers, user_agent_token, disallowed_header_directives):
     return False
 
 
+class RobotFileParser(urllib.robotparser.RobotFileParser):
+    """Extends RobotFileParser to handle Request urls, and to parse responses more in line with RFC9309"""
+
+    def __init__(self, url=""):
+        super().__init__()
+        self.set_url(url)
+
+    def set_url(self, url):
+        """Sets the URL referring to a robots.txt file, from either a url string or a Request object"""
+        self.url = url
+        if isinstance(url, urllib.request.Request):
+            self.host = url.host
+            self.path = url.selector
+        else:
+            self.host, self.path = urllib.parse.urlparse(url)[1:3]
+
+    def read(self):
+        """Reads the robots.txt URL and feeds it to the parser."""
+        try:
+            with urllib.request.urlopen(self.url, timeout=10) as f:
+                if f.info().get_content_type() != "text/plain":
+                    # server did not respond with a parsable robots.txt file, handle the same as unavailable status
+                    self.allow_all = True
+                else:
+                    raw = f.read()
+                    self.parse(raw.decode("utf-8").splitlines())
+        except urllib.error.HTTPError as err:
+            if err.code in (401, 403) or 500 <= err.code <= 599:
+                # server is unreachable, deny all
+                self.disallow_all = True
+            elif 400 <= err.code <= 499:
+                # robots.txt is unavailable, allow all (even though future requests are likely to fail)
+                self.allow_all = True
+        except Exception:  # pylint: disable=broad-except
+            # treat other errors as meaning the server is unreachable (timeout, ssl error, dns failure)
+            self.disallow_all = True
```

Comment:
this should be measured, probably not cached, it is not clear one failure of the server means the server is fully down

Reply:
Most of the errors that lead to this block are network errors (name resolution failures, timeout). There's definitely improvements that can be made for network errors but for this initial implementation I've just gone with the required behaviour described (by MUST) in 2.3.1.4.

Another option could be to let these Exceptions propagate up to be caught by the download_image try/except block, this would skip requesting the current image and try fetching robots.txt the next time the host is encountered in the thread.